### DISCUSSION TITLE
Check php version in composer.json and maybe adjust ddev version accordingly

### DISCUSSION THREAD

**Discussion by kaystrobach** [ddev/ddev#3788] [Ideas]
Composer.json can provide the needed php version.

it would be awesome to fetch this one during start and inform the user if they both do not match.

**Comment by rfay** [comment]
It's got potential! 

I guess it's something like
```json
"config": {
    "platform": {
        "php": "8.0.1"
    }
}
``` 

https://getcomposer.org/doc/06-config.md#platform

It looks like you can use the major version (8.0) instead of patch (8.0.1)

**Reply by nico-loeber** [reply]
Please note that config.platform.php shouldn't be set for most projects or libs, as this is only possible if you know the exact patch level of the environment you deploy to. Mostly you use this setting if your local php version differs from the production one which is already a problem you should solve.

You can parse the required PHP Version from `require.php`

```
{
  "type": "project",
  "license": "proprietary",
  "minimum-stability": "stable",
  "prefer-stable": true,
  "require": {
    "php": "~8.1.0",
    "ext-ctype": "*",
    "ext-iconv": "*",
    "doctrine/orm": "^2.10",
    "symfony/asset": "6.0.*",
     ...
  }
}
```

Please note that  `"php": "~5.6.7|^8.0.22|7.1.15"` is a totally valid value. So I guess you won't do the parsing on your own.
For my ddev integration plugin I had the same idea, then I decided just to use the configured php language level, as PHPStorm does already parse the php version constraint in the project composer.json. ;)

**Reply by jonaseberle** [reply]
> It looks like you can use the major version (8.0) instead of patch (8.0.1)

In `config.platform.php` you should only use real, complete versions. This will be used to "fake" a PHP version when resolving dependencies. This IMHO does only make sense if you are running a different PHP version locally/in CI compared to the live system. This is a problem that we do not have with ddev!

`require.php` on the other hand allows for the usual composer constraints and fails hard when the requirement is not met.

In my opinion we already have everything: We (`composer install`) fail hard if `require.php` is not met - so DDEV does not have to do anything else.

I would advise against using `config.platform.php` except for very (!) specific use cases. It is rather a recipe for failure than anything that makes your setup more stable IMHO.
